examples/common/fdiff_loss.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Monitor rate of change of loss."""
+from __future__ import annotations
+
+from typing import Any, Dict
+
+from composer.core import Callback, State
+from composer.loggers import Logger
+
+
+class FDiffLoss(Callback):
+    """Rate of chage of loss.
+
+    tracks and plots the rate of change of loss effectively taking the numerical
+    derivative of the loss
+    """
+
+    def __init__(self):
+        self.train_prev_loss = None
+        self.train_prev_metric = {}
+        self.eval_prev_metric = {}
+
+    def state_dict(self) -> Dict[str, Any]:
+        return {
+            'train_prev_loss': self.train_prev_loss,
+            'train_prev_metric': self.train_prev_metric,
+            'eval_prev_metric': self.eval_prev_metric,
+        }
+
+    def load_state_dict(self, state: Dict[str, Any]) -> None:
+        self.train_prev_loss = state['train_prev_loss']
+        self.train_prev_metric = state['train_prev_metric']
+        self.eval_prev_metric = state['eval_prev_metric']
```

Comment:
I would strongly consider how much value it adds to have state_dict and resumption (saves you from waiting 1 train batch or 1 eval period).

Because if anything about this callback's state_dict ever changes, it would break checkpoint compatibility. So we would probably be committing to make all future versions of this callback backwards-compatible.

(fyi this is why we ripped almost everything out of `SpeedMonitor.state_dict`. Stateless = no bugs :)

Reply:
I guess if we restart, we'll only be missing one itr idx in the fdiff graph; it wont even be noticeable.
Ok going to make it stateless.